README.md
# [GoLog](https://gitlab.com/GCSBOSS/golog)

A light-weight stdout logging library for NodeJS and [Deno](https://deno.land). Checkout the main features:

- JSON output by default
- Automatic readable output format for dev environments
- Auto-format error, request and repsonse objects.
- Support for adding custom function to parse interesting input data
- `printf`-like formatting on messages.
- Automatic pid field on log entries.
- Log entries filtering by level or type (whitelist, blacklist)

## Get Started NodeJS

1. Install with: `npm i -P golog`.
2. Setup a logger:

```js
const GoLog = require('require');

// Crate a new logger.
var log = new GoLog();
```

## Get Started [Deno](https://deno.land)

```js
import GoLog from 'https://deno.land/x/golog@v0.4.0-0/deno/main.js';

// Crate a new logger.
var log = new GoLog();
```

3. Create new log entries with one of the supported levels:

```js
log.debug('Main log message');
log.info('Main log message');
log.warn('Counting, %d, %d, %d... %s', 1, 2, 3, 'Foo!');
log.error('Main log message');
log.fatal({ merge: 'this', object: 'with', the: 'entry' }, 'The MSG');
```

## Other Features

You can set default properties to show up in every log entry for a given logge rinstance.
For exemple we can use it to set a `source`.

```js
var l1 = new GoLog({ defaults: { source: 'app-a' } });
l1.warn('Hey!');
var l2 = new GoLog({ defaults: { source: 'app-b' } });
l2.warn('Ho!');
```

Default loggers log entries of any level. Change the level for warn or higher.

```js
var log = new GoLog({ level: 'warn' });
log.debug('Let\'s go!');
```

You can check/use log entries right off the bat.

```js
var log = new GoLog();

let e1 = log.fatal('foo');
let e2 = log.debug('bar');
console.log(e1, e2);
```

If you want to log some information about an `Error` instance, just add it as a key to the data object.

```js
log.warn({ err: new Error('My error') }, 'The message lives on');
```

GoLog also formats http request objects.

```js
const http = require('http');
const GoLog = require('require');

var log = new GoLog();

var server = http.createServer(function(req, res){
    res.end();
    log.warn({ req }, 'The message lives on');
}).listen(8765);

http.get('http://localhost:8765');
```

You can create a disabled logger if you send `false` as the input parameter

```js
var log = new GoLog(false);
log.info('Wont log this');
```

## Reporting Bugs
If you have found any problems with this module, please:

1. [Open an issue](https://gitlab.com/GCSBOSS/golog/issues/new).
2. Describe what happened and how.
3. Also in the issue text, reference the label `~bug`.

We will make sure to take a look when time allows us.

## Proposing Features
If you wish to get that awesome feature or have some advice for us, please:
1. [Open an issue](https://gitlab.com/GCSBOSS/golog/issues/new).
2. Describe your ideas.
3. Also in the issue text, reference the label `~proposal`.

## Contributing
If you have spotted any enhancements to be made and is willing to get your hands
dirty about it, fork us and
[submit your merge request](https://gitlab.com/GCSBOSS/golog/merge_requests/new)
so we can collaborate effectively.


CHANGELOG.md
# GoLog Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.5.0] - 2021-12-16

### Fixed
- conf reading to not modify input object

### Removed
- req and res automatic parsers

### Changed
- Deno: interface to have proper TS types

## [v0.4.1] - 2021-03-15

### Added
- hostname key to all log entries

### Fixed
- Deno: bug when receiving non-strings as replacement values

## [v0.4.0] - 2021-01-20

### Changed
- `addParser` interface to receive a `key` name to be transformed

## [v0.3.1] - 2020-08-26

### Fixed
- bug parsing `req` object to consider `IncomingMessage` instead of `ClientRequest`

## [v0.3.0] - 2020-08-26

### Added
- `defaults` option for setting fields to appear on every entry

### Removed
- `app` logger configuration

## [v0.2.0] - 2020-08-23

### Added
- log coloring
- auto-parsing of specific fields
- function to filter out by log level
- function to filter out by log type

### Changed
- entire interface for logging to stdout

### Removed
- support for direct file/stream writing

## [v0.1.2] - 2019-10-15

### Fixed
- 'err' log entry stack array to not contain ending empty string
- file stream creation to append instead of overwrite

## [v0.1.1] - 2019-09-26

### Fixed
- `error` event to not throw an exception when not handled

## [v0.1.0] - 2019-09-25
- First officially published version.

[v0.1.0]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.1.0
[v0.1.1]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.1.1
[v0.1.2]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.1.2
[v0.2.0]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.2.0
[v0.3.0]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.3.0
[v0.3.1]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.3.1
[v0.4.0]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.4.0
[v0.4.1]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.4.1
[v0.5.0]: https://gitlab.com/GCSBOSS/golog/-/tags/v0.5.0


deno/test.js

import Logger from './main.ts'
import { assert, assertEquals } from 'https://deno.land/std/testing/asserts.ts'

Deno.env.set('ENV', 'testing')

Deno.test('Logging - Should log appropriate objects according to log level', function(){
    const log = new Logger()
    assertEquals(log.debug().level, 'debug')
    assertEquals(log.info().level, 'info')
    assertEquals(log.warn().level, 'warn')
    assertEquals(log.error().level, 'error')
    assertEquals(log.fatal().level, 'fatal')
})

Deno.test('Logging - Should printf format message with input arguments', function(){
    assertEquals(new Logger().debug('a b %s d %s', 'c', 'e').msg, 'a b c d e')
})

Deno.test('Logging - Should assign first object argument properties to final log entry', function(){
    const e = new Logger().info({ a: 1, b: 2 })
    assertEquals(e.a, 1)
    assertEquals(e.b, 2)
})

Deno.test('Logging - Should include entry type', function(){
    const log = new Logger()
    assertEquals(log.warn().type, 'event')
    assertEquals(log.error({ type: 'foobar' }).type, 'foobar')
})

Deno.test('Auto-Parsing - Should parse error objects', function(){
    const log = new Logger()
    assert(Array.isArray(log.warn({ err: new Error('Foobar') }).stack))
    assert(Array.isArray(log.info({ err: 'My Error' }).stack))
})

Deno.test('Auto-Parsing - Should execute added parsers', function(){
    const log = new Logger()
    log.addParser('foo', () => ({ foobar: true }))
    const e = log.info({ foo: {} })
    assert(e.foobar)
})

Deno.test('Options - Should not log when entry level is below conf level [level]', function(){
    assert(!new Logger({ level: 'error' }).warn())
})

Deno.test('Options - Should set defaults values for all entries [defaults]', function(){
    const e = new Logger({ defaults: { a: 'foo', b: 'bar' } }).info({ a: 1 })
    assertEquals(e.a, 1)
    assertEquals(e.b, 'bar')
})

Deno.test('Options - Should not log anything when conf is FALSE', function(){
    assert(!new Logger(false).debug())
})

Deno.test('Options - Should only log selected types [only]', function(){
    const log = new Logger({ only: 'a' })
    assert(log.debug({ type: 'a' }))
    assert(!log.debug({ type: 'b' }))
})

Deno.test('Options - Should not log filtered types [except]', function(){
    const log = new Logger({ except: [ 'a', 'c' ] })
    assert(!log.debug({ type: 'a' }))
    assert(log.debug({ type: 'b' }))
    assert(!log.debug({ type: 'c' }))
})


test/spec.js
const assert = require('assert');
const Logger = require('../lib/main');

process.env.NODE_ENV = 'testing';

describe('Logging', function(){
    var log;

    before(function(){
        log = new Logger();
    });

    it('Should log appropriate objects according to log level', function(){
        assert.strictEqual(log.debug().level, 'debug');
        assert.strictEqual(log.info().level, 'info');
        assert.strictEqual(log.warn().level, 'warn');
        assert.strictEqual(log.error().level, 'error');
        assert.strictEqual(log.fatal().level, 'fatal');
    });

    it('Should printf format message with input arguments', function(){
        assert.strictEqual(log.debug('a b %s d %s', 'c', 'e').msg, 'a b c d e');
    });

    it('Should assign first object argument properties to final log entry', function(){
        let e = log.info({ a: 1, b: 2 });
        assert.strictEqual(e.a, 1);
        assert.strictEqual(e.b, 2);
    });

    it('Should include entry type', function(){
        assert.strictEqual(log.warn().type, 'event');
        assert.strictEqual(log.error({ type: 'foobar' }).type, 'foobar');
    });

});

describe('Auto-Parsing', function(){

    it('Should parse error objects', function(){
        let log = new Logger();
        assert(Array.isArray(log.warn({ err: new Error('Foobar') }).stack));
        assert(Array.isArray(log.info({ err: 'My Error' }).stack));
    });

    it('Should execute added parsers', function(){
        let log = new Logger();
        log.addParser('thing', () => ({ foobar: true }));
        let e = log.info({ thing: {} });
        assert(e.foobar);
    });

});

describe('Options', function(){

    it('Should not log when entry level is below conf level [level]', function(){
        let log = new Logger({ level: 'error' });
        assert(!log.warn());
    });

    it('Should set defaults values for all entries [defaults]', function(){
        let log = new Logger({ defaults: { a: 'foo', b: 'bar' } });
        let e = log.info({ a: 1 });
        assert.strictEqual(e.a, 1);
        assert.strictEqual(e.b, 'bar');
    });

    it('Should not log anything when conf is FALSE', function(){
        let log = new Logger(false);
        assert(!log.debug());
    });

    it('Should only log selected types [only]', function(){
        let log = new Logger({ only: 'a' });
        assert(log.debug({ type: 'a' }));
        assert(!log.debug({ type: 'b' }));
    });

    it('Should not log filtered types [except]', function(){
        let log = new Logger({ except: [ 'a', 'c' ] });
        assert(!log.debug({ type: 'a' }));
        assert(log.debug({ type: 'b' }));
        assert(!log.debug({ type: 'c' }));
    });

});

describe('Regressions', function(){

    it('Should not modify the input config', function(){
        const conf = { only: [ 'a' ] };
        new Logger(conf);
        assert(!(conf.only instanceof Set));
    });

});


lib/main.js
const util = require('util');
const os = require('os');

const LEVELS = {
    debug: { w: 0, c: '\x1b[30;5;1m' },
    info: { w: 1, c: '\x1b[0m' },
    warn: { w: 2, c: '\x1b[33m' },
    error: { w: 3, c: '\x1b[31m' },
    fatal: { w: 4, c: '\x1b[41;5;1m' }
};

function output(entry){
    let env = process.env.NODE_ENV;

    // Output friendly log for dev or undfined env
    /* istanbul ignore next */
    if(!env || env === 'development'){
        let uword = (entry.type == 'event' ? entry.level : entry.type).toUpperCase();
        entry.time.setMinutes(entry.time.getMinutes() - entry.time.getTimezoneOffset());
        let utime = entry.time.toISOString().slice(11, -5);
        console.log(LEVELS[entry.level].c + utime + ': ' + uword + ' - ' + entry.msg + '\x1b[0m');
    }

    // Output complete JSON log for production and staging
    /* istanbul ignore next */
    else if(env !== 'testing')
        console.log(JSON.stringify(entry));
}

function parseErr({ err }){
    if(!(err instanceof Error))
        err = new Error(err);

    let stack = err.stack.split(/[\r\n]+\s*/g);
    let env = process.env.NODE_ENV;

    return {
        err: null,
        code: err.code,
        class: err.constructor.name,
        message: err.message,
        stack: stack.slice(1, -1),
        msg: !env || env === 'development'
            ? /* istanbul ignore next */ err.stack
            : stack[0] + ' ' + stack[1]
    };
}

function getEntry(level, args){
    let data = typeof args[0] == 'object' ? args.shift() : {};
    let msg = util.format(...args);
    let type = data.type || 'event';

    /* istanbul ignore next */
    let pid = process.pid != 1 ? process.pid : null;

    for(let key in this.parsers)
        key in data && Object.assign(data, this.parsers[key](data));

    msg = msg || data.msg;
    return { level, type, ...data, msg, pid, time: new Date(),
        hostname: this.hostname };
}

function log(level, ...args){
    let entry = {
        ...this.defaults,
        ...getEntry.call(this, level, args)
    };

    let badLevel = LEVELS[this.level].w > LEVELS[level].w;
    let badType = this.except.has(entry.type) ||
        this.only.size > 0 && !this.only.has(entry.type);
    if(badType || badLevel)
        return false;

    output(entry);

    return entry;
}

module.exports = class GoLog {

    constructor(conf = {}){
        this.level = conf.level || 'debug';
        this.only = new Set([].concat(conf.only).filter(a => a));
        this.except = new Set([].concat(conf.except).filter(a => a));
        this.defaults = conf.defaults || {};
        this.parsers = { err: parseErr };
        this.hostname = os.hostname();

        let op = conf === false ? () => false : log;

        for(let l in LEVELS)
            this[l] = op.bind(this, l);
    }

    addParser(key, parser){
        this.parsers[key] = parser;
    }

};


deno/main.ts
import { sprintf } from 'https://deno.land/std@0.115.0/fmt/printf.ts'

type LevelSpec = { w: number, c: string }
type Dict = Record<string, unknown>

type BasicEntry = {
    time: Date,
    level: string,
    type: string,
    msg: string,
    pid: number | null
}

type ParserFunc = (input: Dict) => Dict

const LEVELS: Record<string, LevelSpec> = {
    debug: { w: 0, c: '\x1b[30;5;1m' },
    info: { w: 1, c: '\x1b[0m' },
    warn: { w: 2, c: '\x1b[33m' },
    error: { w: 3, c: '\x1b[31m' },
    fatal: { w: 4, c: '\x1b[41;5;1m' }
};

function output(entry: Dict){
    const env = Deno.env.get('ENV')

    // Output friendly log for dev or undfined env
    /* istanbul ignore next */
    if(!env || env === 'development'){
        const e = entry as BasicEntry;
        const uword = (e.type == 'event' ? e.level : e.type).toUpperCase()
        e.time.setMinutes(e.time.getMinutes() - e.time.getTimezoneOffset())
        const utime = e.time.toISOString().slice(11, -5)
        console.log(LEVELS[e.level].c + utime + ': ' + uword + ' - ' + e.msg + '\x1b[0m')
    }

    // Output complete JSON log for production and staging
    /* istanbul ignore next */
    else if(env !== 'testing')
        console.log(JSON.stringify(entry))
}

function parseErr({ err }: Dict){
    let theErr: Error;

    if(err instanceof Error)
        theErr = err;
    else
        theErr = new Error(err as string)

    const stack = theErr.stack?.split(/[\r\n]+\s*/g)
    const env = Deno.env.get('ENV')

    return {
        err: null,
        // code: theErr.code,
        class: theErr.constructor.name,
        message: theErr.message,
        stack: stack?.slice(1, -1),
        msg: !env || env === 'development'
            ? theErr.stack
            : stack?.[0] + ' ' + stack?.[1]
    }
}

type GoLogConf = {
    defaults?: Dict,
    level?: string,
    only?: string | string[],
    except?: string | string[]
}

export default class GoLog {

    parsers: Record<string, ParserFunc> = { err: parseErr }
    private level: string
    private only: Set<string>
    private except: Set<string>
    private defaults: Dict

    private log(level: string, ...args: unknown[]){

        const entry = {
            ...this.defaults,
            ...this.getEntry.call(this, level, args)
        }

        const badLevel = LEVELS[this.level].w > LEVELS[level].w
        const badType = this.except.has(entry.type as string) ||
            this.only.size > 0 && !this.only.has(entry.type as string)
        if(badType || badLevel)
            return false

        output(entry)

        return entry
    }

    debug(...args: unknown[]){ return this.log('debug', ...args) }
    info(...args: unknown[]){ return this.log('info', ...args) }
    warn(...args: unknown[]){ return this.log('warn', ...args) }
    error(...args: unknown[]){ return this.log('error', ...args) }
    fatal(...args: unknown[]){ return this.log('fatal', ...args) }

    constructor(conf: false | GoLogConf = {}){
        if(!conf)
            conf = { only: 'badbadclass' }
        this.level = conf.level || 'debug'
        this.only = new Set([ conf.only ].flat().filter((a): a is string => Boolean(a)))
        this.except = new Set([ conf.except ].flat().filter((a): a is string => Boolean(a)))
        this.defaults = conf.defaults || {}
    }

    addParser(key: string, parser: ParserFunc){
        this.parsers[key] = parser
    }

    private getEntry(level: string, args: unknown[]): BasicEntry{
        const data = typeof args[0] == 'object' ? args.shift() as Dict : {}
        args[0] = args[0] || ''
        let msg = sprintf(args[0] as string, ...args.slice(1).map(a => String(a)))

        const type = data.type as string || 'event'

        /* istanbul ignore next */
        const pid = Deno.pid != 1 ? Deno.pid : null

        for(const key in this.parsers)
            key in data && Object.assign(data, this.parsers[key](data))

        msg = msg || String(data.msg);
        return { level, type, ...data, msg, pid, time: new Date() }
    }

}
